onnxruntime/core/optimizer/shape_to_initializer.cc

```diff
@@ -0,0 +1,82 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "core/optimizer/shape_to_initializer.h"
+#include "core/graph/graph.h"
+#include "core/graph/graph_utils.h"
+#include "core/graph/op.h"
+#include "core/optimizer/initializer.h"
+#include "core/optimizer/optimizer_execution_frame.h"
+#include "core/framework/op_kernel.h"
+#include "core/framework/tensorprotoutils.h"
+
+namespace onnxruntime {
+
+Status ShapeToInitializer::Apply(Graph& graph, Node& node, RewriteRuleEffect& rule_effect) const {
+  // Create an execution frame to get access to the memory allocator.
+  OptimizerExecutionFrame::Info info({&node}, InitializedTensorSet());
+
+  // Store the statically inferred shape of the input to the Shape operator.
+  const ONNX_NAMESPACE::TensorShapeProto* input_shape_proto = node.InputDefs()[0]->Shape();
+  std::vector<int64_t> input_dims;
+  int num_dimensions = input_shape_proto->dim_size();
+  for (int i = 0; i < num_dimensions; i++) {
+    input_dims.push_back(gsl::narrow_cast<int64_t>(input_shape_proto->dim(i).dim_value()));
+  }
+
+  // Create the TensorProto that will be used as initializer in place of the Shape operator.
+  const auto* shape_out_def = node.OutputDefs()[0];
+
+  ONNX_NAMESPACE::TensorProto shape_initializer_proto;
+
+  shape_initializer_proto.set_name(shape_out_def->Name());
+
+  TensorShape tensor_shape({gsl::narrow_cast<int64_t>(num_dimensions)});
+  for (auto& dim : tensor_shape.GetDims()) {
+    shape_initializer_proto.add_dims(dim);
+  }
+
+  auto tensor_proto_data_type = shape_out_def->TypeAsProto()->tensor_type().elem_type();
+
+  shape_initializer_proto.set_data_type(tensor_proto_data_type);
+
+  // Here we expect little-indian format to set raw data of the TensorProto.
+  shape_initializer_proto.set_raw_data(input_dims.data(),
+                                       input_dims.size() * sizeof(decltype(input_dims)::value_type));
+
+  // Remove the output edges of the Shape node, then remove the node itself, and replace it with the initializer.
+  graph_utils::RemoveNodeOutputEdges(graph, node);
+
+  if (graph.RemoveNode(node.Index())) {
+    rule_effect = RewriteRuleEffect::kRemovedCurrentNode;
+  }
+
+  graph.AddInitializedTensor(shape_initializer_proto);
+
```

Comment:
If remove node failed do you still want to do this?

Reply:
I moved it inside the remove-node if (it was okay since we would do the initializer cleanup, but it's better this way), thanks Scott.